casper/src/main/scala/coop/rchain/casper/merging/MergeScope.scala

```diff
@@ -0,0 +1,195 @@
+package coop.rchain.casper.merging
+
+import cats.Applicative
+import cats.effect.Concurrent
+import cats.syntax.all._
+import com.google.protobuf.ByteString
+import coop.rchain.blockstorage.dag.{Finalizer, Message}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.FringeData
+import coop.rchain.models.Validator.Validator
+import coop.rchain.models.syntax._
+import coop.rchain.rholang.interpreter.RhoRuntime.RhoHistoryRepository
+import coop.rchain.rholang.interpreter.merging.RholangMergingLogic
+import coop.rchain.rholang.syntax._
+import coop.rchain.rspace.HotStoreTrieAction
+import coop.rchain.rspace.hashing.Blake2b256Hash
+import coop.rchain.rspace.merger.EventLogMergingLogic.NumberChannelsDiff
+import coop.rchain.rspace.merger.{ChannelChange, StateChange, StateChangeMerger}
+import coop.rchain.rspace.syntax._
+import coop.rchain.sdk.dag.merging.ConflictResolutionLogic
+import coop.rchain.sdk.dag.merging.ConflictResolutionLogic.{conflictScope, finalScope}
+import coop.rchain.sdk.syntax.all._
+import coop.rchain.shared.{Log, Stopwatch}
+import scodec.bits.ByteVector
+
+/**
+  * The scope to merge.
+  * @param finalScope     blocks that cannot be altered
+  * @param conflictScope  blocks that can be altered
+  * @param baseState      base state to merge into
+  */
+final case class MergeScope(
+    finalScope: Set[BlockHash],
+    conflictScope: Set[BlockHash],
+    baseState: Blake2b256Hash
+)
+
+object MergeScope {
+
+  private def findGenesis(
+      dagData: Map[BlockHash, Message[BlockHash, Validator]]
+  ): Option[BlockHash] = {
+    val candidates = dagData.filter {
+      case (_, m) =>
+        val edge      = m.seen == Set(m.id)
+        val seenByAll = dagData.valuesIterator.forall(_.seen.contains(m.id))
+        edge && seenByAll
+    }
+    if (candidates.size > 1) none[BlockHash] else candidates.headOption.map(_._1)
+  }
+
+  /**
+    * Use dag data to optimise the merge scope.
+    * For now covers only the case of genesis ceremony, when final set is empty so
+    * genesis block can be removed from conflict set and merge can be done into genesis post state.
+    */
+  def optimise[F[_]: Applicative](
```

Comment:
This is the place where DAG structure can be used to optimise Merge Scope. After this point DAG is forgotten.